[FSDP2] Introduced initial `fully_shard` frontend

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* #117861
* #117852
* #117817
* #117814
* __->__ #117776
* #117984
* #118186
* #117994

This PR introduces the initial `fully_shard` frontend without any distributed logic that will be built into per-parameter-sharding FSDP.
- We design `fully_shard` to be a _module-level_ API (taking in an `nn.Module`), e.g. as opposed to a tensor-level one.
- We define a `FSDP` class and use a dynamic class swap, setting `module.__class__` to a newly created class that subclasses `FSDP` and `type(module)`, to allow FSDP to override and add methods on the module.
    - We name this class as `FSDP<type(module)>`, e.g. `FSDPLinear` for `Linear`.
    - We disable the `deepcopy` because the state object inserted on the module will not be trivially `deepcopy`-able.
- Calling `fully_shard(module)` inserts a state object on `module` but not any of its children. This state object will be used for any FSDP-specific state.
- We raise an error on `ModuleList` or `ModuleDict` since they do not implement `forward()`, and FSDP will rely on `forward()` to insert logic (https://github.com/pytorch/pytorch/issues/113794).
- In the future, we will deprecate the existing `fully_shard` that calls into the same backend logic as `FullyShardedDataParallel` as there is no adoption for that and we prefer to reuse that name.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[FSDP2] Introduced initial `fully_shard` frontend

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Introduced initial `fully_shard` frontend"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,44 @@
+    )
+
+
+class FSDP:
```
is FSDP a convenience for people who are used to current FSDP to migrate? do we want this?  (why not just tell them use fully_shard)

awgu replied:
This class is used for us to add and override methods (see the PR description for explanation). It is distinct from `fully_shard` the frotend.

awgu replied:
Without a class like `FSDP`, we would have to monkey patch to add or override methods. Now, we can just define them on `FSDP` and since `FSDP` is earliest in the method resolution order, it will override any existing method definition.

wconstab replied:
nvm i see that its used mainly as a target for isinstance checks.

mby worth documenting that it's not intended for manual use but is the class that your fully_shard(module)  takes on when sharded

wconstab replied:
oh, your explanation helped even more. didn't see it when i typed the above. thanks!

wconstab commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,44 @@
+    cls = module.__class__
+    dct = {"__deepcopy__": unimplemented_deepcopy}
+    new_cls = type(f"FSDP{cls.__name__}", (FSDP, cls), dct)
+    module.__class__ = new_cls
```
@voznesenskym do we no longer care about deepcopying all the modules we dynamo trace? or do we plan to handle FSDP some other way?

wconstab commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,44 @@
+
+
+@contract()
+def fully_shard(
```
should definitely apidoc this one

awgu replied:
Does it need to be done in this PR? The `fully_shard` function is barebones here and does not really do anything. Since this is not a public API, I was thinking to defer the docstring to when there is actually some semblance of FSDP being implemented.

awgu replied:
I am also happy to fill in the docstring here with some description first though.

wconstab replied:
yea thats fine

wconstab reviewed (approved):
LGTM!

Skylion007 commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,44 @@
+    return module
+
+
+def unimplemented_deepcopy(*args: Any, **kwargs: Any) -> None:
```
This should return typing_extensions.Never

Commit:
Update on "[FSDP2] Introduced initial `fully_shard` frontend"


This PR introduces the initial `fully_shard` frontend without any distributed logic that will be built into per-parameter-sharding FSDP.
- We design `fully_shard` to be a _module-level_ API (taking in an `nn.Module`), e.g. as opposed to a tensor-level one.
- We define a `FSDP` class and use a dynamic class swap, setting `module.__class__` to a newly created class that subclasses `FSDP` and `type(module)`, to allow FSDP to override and add methods on the module.
    - We name this class as `FSDP<type(module)>`, e.g. `FSDPLinear` for `Linear`.
    - We disable the `deepcopy` because the state object inserted on the module will not be trivially `deepcopy`-able.
- Calling `fully_shard(module)` inserts a state object on `module` but not any of its children. This state object will be used for any FSDP-specific state.
- We raise an error on `ModuleList` or `ModuleDict` since they do not implement `forward()`, and FSDP will rely on `forward()` to insert logic (https://github.com/pytorch/pytorch/issues/113794).
- In the future, we will deprecate the existing `fully_shard` that calls into the same backend logic as `FullyShardedDataParallel` as there is no adoption for that and we prefer to reuse that name.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,46 @@
+        raise ValueError(
+            f"fully_shard does not support containers that do not implement forward: {module}"
+        )
+    state = fully_shard.state(module)
```
nit: it took me some time to find out ``fully_shard.state`` is defined in ``@contract``. maybe consider a short comment here to remind people to check ``@contract`` docstrings?

wanchaol commented on test/distributed/_composable/fsdp/_test_fully_shard_common.py:
```diff
@@ -0,0 +1,20 @@
+import torch.nn.functional as F
+
+
+class MLP(nn.Module):
```
it feels to me that the `MLP` module is so popular in testing (dtensor also have [one](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/distributed/_tensor/common_dtensor.py#L53)), maybe as a follow up BE task we should consolidate these two and put it under `common_distributed`?

awgu replied:
Good point! (We will want a buffer on the MLP module though in a later PR 😄 .)

wanchaol commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,46 @@
+    # Place FSDP leftmost for highest priority in the method resolution order
+    cls = module.__class__
+    dct = {"__deepcopy__": unimplemented_deepcopy}
+    new_cls = type(f"FSDP{cls.__name__}", (FSDP, cls), dct)
```
I thought this should be the below instead?
```
f"FSDP<{cls.__name__}>"
```

awgu replied:
For my understanding, is this mainly a stylistic difference of what the new class name is? Or, is there another reason why we want the `<>`?

If we have the `<>`, is that a valid class name? I was not sure how you can do `isinstance(module, FSDP<Linear>)` (if we hypothetically had access to the dynamically constructed class), whereas `isinstance(module, FSDPLinear)` seems syntactically valid.

wanchaol replied:
Oh it's mainly stylistic difference and I was only thinking about sth like `__str__` to nicely show it's a FSDP wrapped module, but the `isinstance` check make sense we should just have a valid class name without `<>`

wanchaol commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -0,0 +1,46 @@
+            f"fully_shard does not support containers that do not implement forward: {module}"
+        )
+    state = fully_shard.state(module)
+    _insert_module_state(module, state)
```
I think it would be useful to put some comment about things like the `module` and `state` relationship (i.e. is it 1-1 or 1-many), the root `state` difference compare with the child states.

wanchaol reviewed (approved):
looks good! only have some minor commnets

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_state.py:
```diff
@@ -0,0 +1,93 @@
+
+from torch.distributed._composable.fsdp import FSDP, fully_shard
+from torch.testing._internal.common_cuda import TEST_CUDA
+from torch.testing._internal.common_fsdp import FSDPTestMultiThread
```
I personally wanted to keep it separate for organization and moving faster, but it is not a dealbreaker either way.
- Organization: The dependency is unidirectional -- per-parameter FSDP tests may use utils from existing FSDP but not vice versa. Keeping separate may help avoid bloating the existing FSDP common file with code that existing FSDP tests never use.
- Moving faster: In case people change the existing FSDP common file (unlikely but possible), I would prefer not to have to manage rebase conflicts through this stack of PRs 😃 .

wanchaol replied:
sounds good yeah!

Commit:
Update on "[FSDP2] Introduced initial `fully_shard` frontend"


This PR introduces the initial `fully_shard` frontend without any distributed logic that will be built into per-parameter-sharding FSDP.
- We design `fully_shard` to be a _module-level_ API (taking in an `nn.Module`), e.g. as opposed to a tensor-level one.
- We define a `FSDP` class and use a dynamic class swap, setting `module.__class__` to a newly created class that subclasses `FSDP` and `type(module)`, to allow FSDP to override and add methods on the module.
    - We name this class as `FSDP<type(module)>`, e.g. `FSDPLinear` for `Linear`.
    - We disable the `deepcopy` because the state object inserted on the module will not be trivially `deepcopy`-able.
- Calling `fully_shard(module)` inserts a state object on `module` but not any of its children. This state object will be used for any FSDP-specific state.
- We raise an error on `ModuleList` or `ModuleDict` since they do not implement `forward()`, and FSDP will rely on `forward()` to insert logic (https://github.com/pytorch/pytorch/issues/113794).
- In the future, we will deprecate the existing `fully_shard` that calls into the same backend logic as `FullyShardedDataParallel` as there is no adoption for that and we prefer to reuse that name.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Introduced initial `fully_shard` frontend"


This PR introduces the initial `fully_shard` frontend without any distributed logic that will be built into per-parameter-sharding FSDP.
- We design `fully_shard` to be a _module-level_ API (taking in an `nn.Module`), e.g. as opposed to a tensor-level one.
- We define a `FSDP` class and use a dynamic class swap, setting `module.__class__` to a newly created class that subclasses `FSDP` and `type(module)`, to allow FSDP to override and add methods on the module.
    - We name this class as `FSDP<type(module)>`, e.g. `FSDPLinear` for `Linear`.
    - We disable the `deepcopy` because the state object inserted on the module will not be trivially `deepcopy`-able.
- Calling `fully_shard(module)` inserts a state object on `module` but not any of its children. This state object will be used for any FSDP-specific state.
- We raise an error on `ModuleList` or `ModuleDict` since they do not implement `forward()`, and FSDP will rely on `forward()` to insert logic (https://github.com/pytorch/pytorch/issues/113794).
- In the future, we will deprecate the existing `fully_shard` that calls into the same backend logic as `FullyShardedDataParallel` as there is no adoption for that and we prefer to reuse that name.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

awgu commented:
I am landing this PR to try to reduce the height of the stack. I am going to try to change others still under review to `[no ci]`.

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @awgu! As we discussed previously, let's move _test_fully_shard_common to torch.testing._internal.common_fsdp.

awgu commented:
I will re-land this as a new PR since I had to rebase my stack that existed on top of this one.
